examples/credential-registry-storage-contract/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+    Expired,
+}
+
+type ContractResult<A> = Result<A, CustomContractError>;
+
+/// Mapping the logging errors to CustomContractError.
+impl From<LogError> for CustomContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// Credential state that is used as the input parameter or the return value for
+/// the `store/view` contract functions. It contains the metadata and the
+/// credential_secret.
+#[derive(Serialize, SchemaType, PartialEq)]
+struct SimpleCredentialState {
+    /// Metadata associated with the credential.
+    metadata:          Metadata,
+    /// The credential_secret.
+    credential_secret: Vec<u8>,
+}
+
+/// The CredentialRegisteredEvent is logged when the `store` function is
+/// invoked.
+#[derive(Serialize, SchemaType)]
+pub struct CredentialRegisteredEvent {
+    /// Public_key associated with the stored credential.
+    public_key: PublicKeyEd25519,
+}
+
+/// Tagged event to be serialized for the event log.
+#[derive(SchemaType, Serialize)]
+pub enum Event {
+    CredentialRegistered(CredentialRegisteredEvent),
+}
+
+// Contract functions
+
+/// Initialize a new contract instance.
+#[init(contract = "credential-registry-storage", event = "Event")]
+fn contract_init<S: HasStateApi>(
+    _ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    Ok(State::empty(state_builder))
+}
+
+/// View the registered credential data.
+/// It rejects if:
+/// - It fails to parse the parameter.
+#[receive(
+    contract = "credential-registry-storage",
+    name = "view",
+    parameter = "PublicKeyEd25519",
+    return_value = "SimpleCredentialState",
+    error = "CustomContractError"
+)]
+fn view<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<Option<SimpleCredentialState>> {
+    let public_key: PublicKeyEd25519 = ctx.parameter_cursor().get()?;
+
+    Ok(host.state().credential_registry.get(&public_key).map(|credential_state| {
+        SimpleCredentialState {
+            metadata:          credential_state.metadata,
+            credential_secret: credential_state.credential_secret.clone(),
+        }
+    }))
+}
+
+/// Part of the parameter type for the contract function `store`.
+/// Specifies the message that is signed.
+#[derive(SchemaType, Serialize)]
+struct SignedMessage {
+    /// The contract_address that the signature is intended for.
+    contract_address:  ContractAddress,
+    /// A timestamp to make signatures expire.
+    timestamp:         Timestamp,
+    /// The entry_point that the signature is intended for.
+    entry_point:       OwnedEntrypointName,
+    /// Metadata associated with the credential.
+    metadata:          Metadata,
+    /// The serialized credential_secret.
+    #[concordium(size_length = 2)]
+    credential_secret: Vec<u8>,
+}
```

Comment:
I am not sure this is needed here. The idea is that the keys are separate anyhow and as I though we just sign the message prefixed with
```
WEB3ID:STORE
```

Do you think it is necessary to add this context to the message?

Reply:
This prefixed seems `arbitrary`. Meaning if third parties clone this smart contract and use it for their own use case, they also come up with `arbitrary` prefixes, and one day there are two smart contracts with the same prefix and the same input parameter length and you could use one signature on both contracts. I would suggest adding the `contract_address` to the signed message to avoid that problem. `Timestamp` is discussable (it is a nice feature to have expired signatures). `entry_point` is needless since we only have one entrypoint in this contract (we can remove it or keep it if we want to show how to write a contract which has several signature verification entrypoints).